[Feature] Payment, Shipping, and Billing Information Form Validation

j-bunk opened the pull request:
## Notion ticket link
<!-- Please replace with your ticket's URL -->
[Payment, Shipping, and Billing Information Form Validation](https://www.notion.so/uwblueprintexecs/Payment-Shipping-and-Billing-Information-Form-Validation-4b6cb561e1c1468b8715cad721d000b1)


<!-- Give a quick summary of the implementation details, provide design justifications if necessary -->
## Implementation description
Tried to separate the changes/logic into distinct commits to make it easier to review
* Replaced form fields with Formik wrapped input fields https://github.com/uwblueprint/richmond-centre-for-disability/pull/163/commits/57d2dbea076e84ae18e154460923c93d5eab1375
* Integrated formik into `EditModal` to work with the modified form https://github.com/uwblueprint/richmond-centre-for-disability/pull/163/commits/28c4ad190cce702722f95a8428770dff76db5969
* Integrated formik into `create-new/replacement/renewal/replacement` to work with the modified form https://github.com/uwblueprint/richmond-centre-for-disability/pull/163/commits/41c122644c0e38f6280d313a9957a3adbba023cf
* Created Yup validation schema https://github.com/uwblueprint/richmond-centre-for-disability/pull/163/commits/7e2dc296e00ce4a5caf6161cd4835b46f250886c

<!-- Catch all section that could be used to draw attention to anything the reviewers should keep in mind, substantial parts of your PR, anything you'd like a second opinion on, things that will be fixed in the future, or things that were left out -->
## Notes
* I didn't implement backend validation yet, I think I'll address that separately and start on another form instead

## Screenshots
<img width="591" alt="image" src="https://user-images.githubusercontent.com/51224641/160722350-b9588958-f79a-4b25-8422-650d650ecba5.png">
<img width="901" alt="image" src="https://user-images.githubusercontent.com/51224641/160722611-e18778cb-0352-40f2-b456-715c8a880202.png">

## Checklist
- [x] My PR name is descriptive, is in imperative tense and starts with one of the following: `[Feature]`,`[Improvement]` or `[Fix]`,
- [x] I have run the appropriate linter(s)
- [x] I have requested a review from the RCD team on GitHub, or specific people who are associated with this ticket

j-bunk commented on components/admin/requests/payment-information/EditModal.tsx:
```diff
@@ -22,79 +24,16 @@ type EditPaymentDetailsModalProps = {
-      billingPostalCode,
-    });
+  const handleSubmit = (values: { paymentInformation: PaymentInformationFormData }) => {
+    // TODO FIX: Card is passing in processingFee in object which is breaking request and needs to be removed
+    const { processingFee, ...pickedValues } = values.paymentInformation;
```
I'll look into this tonight/tomorrow. Didn't want this minor issue to delay reviews so I still requested reviews

angeladietz replied:
I resolved this by manually picking all of the values which is how it was being done before

angeladietz commented on components/form/TextField.tsx:
```diff
@@ -7,22 +7,38 @@ import {
 type Props = InputProps & {
   readonly name: string;
-  readonly label: string;
+  readonly label: string | ReactNode;
```
Hm I came across the same problem with the label and I added an extra prop for a `labelHelper`, it looks like yours is more robust and a bit cleaner so i'll go with this too.

angeladietz commented on components/admin/requests/payment-information/Form.tsx:
```diff
@@ -228,87 +166,69 @@ export default function PaymentDetailsForm({
           paddingBottom="24px"
-          isChecked={billingAddressSameAsHomeAddress}
-          onChange={handleChange('billingAddressSameAsHomeAddress')}
+          // isChecked={paymentInformation.billingAddressSameAsHomeAddress}
```
```suggestion
          // isChecked={paymentInformation.billingAddressSameAsHomeAddress}
```
```suggestion
```

angeladietz replied:
```suggestion
```

angeladietz commented on components/admin/requests/payment-information/Form.tsx:
```diff
@@ -228,87 +166,69 @@ export default function PaymentDetailsForm({
+                required
+                placeholder="Select country"
+              >
+                {/* TODO: Add more countries? */}
```
Hmm true, I think if we don't add more countries rn we maybe should just leave it as an `Input`. @OustanDing what are your thoughts?

angeladietz replied:
Making this a disabled text input for now re discussion with team.

angeladietz commented on components/admin/requests/payment-information/Form.tsx:
```diff
@@ -228,87 +166,69 @@ export default function PaymentDetailsForm({
+                required
+                placeholder="Select country"
+              >
+                {/* TODO: Add more countries? */}
```
Hmm true, I think if we don't add more countries rn we maybe should just leave it as an `Input`. @OustanDing what are your thoughts?

angeladietz commented on components/form/TextField.tsx:
```diff
@@ -7,22 +7,38 @@ import {
       <FormLabel htmlFor={name}>{label}</FormLabel>
-      <Input {...field} {...inputProps} />
+      {monetaryInput ? (
+        <>
```
```suggestion
```

angeladietz replied:
I think you can remove the `<>` and `</>` wrapping the `InputGroup` and `Input` on line 39

angeladietz commented on pages/admin/request/create-new.tsx:
```diff
@@ -302,7 +301,7 @@ export default function CreateNew() {
     }
 
-    if (!paymentDetails.paymentMethod) {
+    if (!values.paymentInformation.paymentMethod) {
```
I think we should be able to remove this check and instead just use the validation schema to validate `values` which will ensure its non-null. For example, I do this [here](https://github.com/uwblueprint/richmond-centre-for-disability/blob/ad10ae7e3330a8c54dd60b163b9d80a52e4b3bf5/pages/admin/request/create-new.tsx#L272). Once we merge eveything together we won't have to implement this twice but depending on how long it takes for the other prs to get merged it might be a good idea to implement it here now anyways to get rid of the explicit checks. I believe the same changes can be made in the other files too.

j-bunk replied:
Curious why we do
https://github.com/uwblueprint/richmond-centre-for-disability/blob/ad10ae7e3330a8c54dd60b163b9d80a52e4b3bf5/pages/admin/request/create-new.tsx#L272 in this file?

Wouldn't any value that reaches this point be already validated since the submit button is disabled if the frontend validation fails? And if we wanted to add backend validation I think it would go in the resolver instead.

angeladietz replied:
Yeah, at this point the validation should never fail. The reason we still call it here is because `.validate()` returns an object with all of the validation constraints enforced in the return type. This means if in our validation schema we require a field to be non-null, that field in the result of `.validate()` will have a non-null type. Without doing this, we get a bunch of linting errors since the original `values` is a type which allows fields to be null, because this is the same type as our initial values in formik.

angeladietz commented on pages/admin/request/create-new.tsx:
```diff
@@ -354,8 +353,8 @@ export default function CreateNew() {
-          ...paymentDetails,
-          paymentMethod: paymentDetails.paymentMethod,
+          ...values.paymentInformation,
+          paymentMethod: values.paymentInformation.paymentMethod,
```
won't need to explicitly set paymentMethod once values are validated with above fix.

j-bunk replied:
Addressed but relates to https://github.com/uwblueprint/richmond-centre-for-disability/pull/163#discussion_r841191472

angeladietz reviewed (changes requested):
Looks great! Just left a few comments, some of which will conflict with the other branches so we'll have to merge very carefully

j-bunk commented:
@angeladietz thanks for the review! Addressed the PR feedback and rebased to resolve the merge conflicts. The conditionally rendered parts of the form disable the button if they're not filled in so I'll have to look into that. I still also have to address this https://github.com/uwblueprint/richmond-centre-for-disability/pull/163#discussion_r837977630

OustanDing reviewed (approved):
LGTM! Great work :D
